extensions/security/runtime/src/main/java/io/quarkus/security/runtime/QuarkusSecurityIdentity.java

```diff
@@ -121,11 +123,16 @@ public static Builder builder() {
     }
 
     public static Builder builder(SecurityIdentity identity) {
-        return new Builder()
+        if (identity.getPrincipal() == null && !identity.isAnonymous()) {
```

Comment:
I would check this condition in the build method and not here. With the builder you can manipulate all parameters and the input must not necessary be a valid SecurityIdentity.

Reply:
@renegrob we can't reset the anonymous flag via the builder API right now - and also, should one expect a consistent `Securityidentity` state at this point ?

Reply:
@renegrob I.e, even if `setAnonymous` is added, let the users start from a clean builder as opposed to pass the invalid `SecurityIdentity` instance (which is not modifiable)

Reply:
@sberyozkin validation should be done at build(), not on a partially completed object, and the builder should have a `setAnonymous` method.

Reply:
@stuartwdouglas there are 2 builder methods, this one takes a final `SecurityIdentity` so I don't agree that a completed `Securityidentity` should be accepted in the incorrect state - here once again we'd rely on the user to follow with the right sequence to fix a broken `Securityidentity`.

> and the builder should have a setAnonymous method.

@renegrob Can you open a PR please ?

Thanks

Reply:
@stuartwdouglas @renegrob I've looked at it again. IMHO, without `setAnonymous` it does make sense to keep the current PR as is for the reasons suggested above (Input SecurityIdentity must be correct). `setAnonymous` would change everything, i.e, makes no sense to check at the start and the end - only at the end.
So I'll just push this check to the end and be done with it - this PR can then be merged for 1.13.0.Final or whenever Quarkus gets updated to use the updated `quarkus-security`

Reply:
@stuartwdouglas @renegrob done - it is there (if we start without `setAnonymous`) or nearly there if not :-)

Reply:
Hey @stuartwdouglas @renegrob Just one more line, not for the sake of continuing the argument but simply to add what I forgot to say earlier to explain better why I'm feeling not excited about dropping the early check: adding this copy constructor was the only contribution I had to `QuarkusSecurityIdentity` - the goal is about avoiding a boilerrplate code the custom `SecurityIdentityAugmentor` has to do in order to add a new attribute or credential type - we used to have all the demos showing a lot of copying from the original final identity which is what this copy constructor has eliminated. Custom `SecurityIdentityAugmentor` enhancing the existing `SecurityIdentity` (which was already prepared by `IdentityProvider` or a fellow higher priority `SecurityIdentityAugmentor`) is not about **fixing** it - it is about enhancing the already completed `SecurityIdentity`. The custom `SecurityIdentityAugmentor` which tries to enhance it does not know the existing state - it only adds something to it - hence the late check will `surprise` this current `SecurityIdentityAugmentor` which only added some custom credential if it happens it accepted an invalid `SecurityIdentity`.
Hence it is not a very sound approach IMHO - but I agree it is not a big deal since the check will be done anyway.
That was the last comment from me, promise :-)

Reply:
Sorry, it was my fault that there was a detour. And I'm just a contributor and not owner of any code. Maybe I should just listen more and comment less. At least for me, I don't have a problem with your comments. They make sense and it is normal to have different point of views. Ideally we find a common agreement which can be different from anyones initial opinion or idea. And my opinion shouldn't have too much weight.

Reply:
@renegrob You are very much welcome with your comments and I give them the same weight as to comments from @stuartwdouglas in this case - you are the author of this issue, have already contributed, so I'm CC-ing you and take what you say on board :-). Yes - we can sometimes have a rather long discussion but it can be worth it, especially, when it comes to the sec issues :-)